integration/users/connection_test.go

```diff
@@ -307,6 +308,181 @@ func TestAuthenticationOnAuthenticatedConnection(t *testing.T) {
+			authenticationDB: "admin",
+			path:             "/",
+		},
+		"DefaultAuthDB": {
+			// example: mongodb://user1:pass1@127.0.0.1:40623/TestAuthenticationAuthSource
+			username:         "user1",
+			password:         "pass1",
+			authenticationDB: t.Name(),
+			path:             t.Name(),
+		},
+		"AuthSource": {
+			// example: mongodb://user2:pass2@127.0.0.1:40623/?authSource=TestAuthenticationAuthSource
+			username:         "user2",
+			password:         "pass2",
+			authenticationDB: t.Name(),
+			path:             "/",
+			authSource:       t.Name(),
+		},
+		"AuthSourceWithDB": {
+			// example: mongodb://user3:pass3@127.0.0.1:40623/XXX?authSource=TestAuthenticationAuthSource
+			username:         "user3",
+			password:         "pass3",
+			authenticationDB: t.Name(),
+			path:             "XXX",
+			authSource:       t.Name(),
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			roles := bson.A{"readWrite"}
+			if !setup.IsMongoDB(t) {
+				// TODO https://github.com/FerretDB/FerretDB/issues/3974
+				roles = bson.A{}
+			}
+
+			err := db.Client().Database(tc.authenticationDB).RunCommand(ctx, bson.D{
+				{"createUser", tc.username},
+				{"roles", roles},
+				{"pwd", tc.password},
+			}).Err()
+			require.NoError(t, err)
+
+			t.Cleanup(func() {
+				err = db.Client().Database(tc.authenticationDB).RunCommand(ctx, bson.D{{"dropUser", tc.username}}).Err()
+				require.NoError(t, err)
+			})
+
+			u, err := url.Parse(s.MongoDBURI)
+			require.NoError(t, err)
+
+			u.Path = tc.path
+			u.User = url.UserPassword(tc.username, tc.password)
+
+			q := u.Query()
+			q.Set("authSource", tc.authSource)
+
+			if tc.authSource == "" {
+				q.Del("authSource")
+			}
+
+			u.RawQuery = q.Encode()
+
+			opts := options.Client().ApplyURI(u.String())
+
+			t.Log("connecting", u.String())
+
+			client, err := mongo.Connect(ctx, opts)
+			require.NoError(t, err)
+
+			t.Cleanup(func() {
+				require.NoError(t, client.Disconnect(ctx))
+			})
+
+			_, err = client.Database(tc.authenticationDB).Collection("test").Find(ctx, bson.D{})
+			require.NoError(t, err)
+		})
+	}
+}
+
+func TestAuthenticationDifferentDatabase(t *testing.T) {
+	t.Parallel()
+
+	s := setup.SetupWithOpts(t, nil)
+	ctx, db := s.Ctx, s.Collection.Database()
+	user, db1, pass1, db2, pass2 := "user", t.Name()+"db1", "pass1", t.Name()+"db2", "pass2"
+
+	for dbName, pass := range map[string]string{
+		db1: pass1,
+		db2: pass2,
+	} {
+		roles := bson.A{"readWrite"}
+		if !setup.IsMongoDB(t) {
+			// TODO https://github.com/FerretDB/FerretDB/issues/3974
+			roles = bson.A{}
+		}
+
+		err := db.Client().Database(dbName).RunCommand(ctx, bson.D{
+			{"createUser", user},
+			{"roles", roles},
+			{"pwd", pass},
+		}).Err()
+		require.NoError(t, err)
+
+		t.Cleanup(func() {
+			err = db.Client().Database(dbName).RunCommand(ctx, bson.D{{"dropAllUsersFromDatabase", int32(1)}}).Err()
+			require.NoError(t, err)
+		})
+	}
+
+	for name, tc := range map[string]struct {
```

Comment:
There is one case that I don't see in this or the previous test.
What if `authSource` is set as something else than a `db1`? 
E.g. the user is correct, the password is correct, but `authSource` is set as a non-existent database, or `authSource` database exists, but no users were created with this database.